src/transport/TCP.h

```diff
@@ -0,0 +1,244 @@
+     *   The class allows separate definitions to allow local execution of several
+     *   Nodes.
+     */
+    CHIP_ERROR Init(TcpListenParameters & params);
+
+    CHIP_ERROR SendMessage(const MessageHeader & header, const PeerAddress & address, System::PacketBuffer * msgBuf) override;
+
+    void Disconnect(const PeerAddress & address) override;
+
+    bool CanSendToPeer(const PeerAddress & address) override
+    {
+        return (mState == State::kInitialized) && (address.GetTransportType() == Type::kTcp) &&
+            (address.GetIPAddress().Type() == mEndpointType);
+    }
+
+    /**
+     * Helper method to determine if IO processing is still required for a TCP transport
+     * before everything is cleaned up (socket closing is async, so after calling 'Close' on
+     * the transport, some time may be needed to actually be able to close.)
+     */
+    bool HasActiveConnections() const;
+
+    /**
+     * Close all active connections.
+     */
+    void CloseActiveConnections();
+
+private:
+    /**
+     * Find an active connection to the given peer or return nullptr if
+     * no active connection exists.
+     */
+    Inet::TCPEndPoint * FindActiveConnection(const PeerAddress & addr);
+
+    /**
+     * Sends the specified message once a connection has been established.
+     *
+     * @param addr - what peer to connect to
+     * @param msg - what buffer to send once a connection has been established.
+     *
+     * Ownership of msg is taken over and will be freed at some unspecified time
+     * in the future (once connection succeeds/fails).
+     */
+    CHIP_ERROR SendAfterConnect(const PeerAddress & addr, System::PacketBuffer * msg);
+
+    /**
+     * Process a single received buffer from the specified peer address.
+     *
+     * @param endPoint the source end point from which the data comes from
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the actual data
+     *
+     * Ownership of buffer is taken over and will be freed (or re-enqueued to the endPoint receive queue)
+     * as needed during processing.
+     */
+    CHIP_ERROR ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress, System::PacketBuffer * buffer);
+
+    /**
+     * Process a single message of the specified size from a buffer.
+     *
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the buffer containing the message
+     * @param messageSize how much of the head contains the actual message
+     *
+     * Does *not* free buffer.
+     */
+    CHIP_ERROR ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
+                                                  uint16_t messageSize);
+
+    // Callback handler for TCPEndPoint. TCP message receive handler.
+    static void OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuffer * buffer);
```

Comment:
This should probably document ownership semantics too...

Reply:
Should be handled by TCPEndPoint. That documents it as:

```
     *  A data reception event handler must acknowledge data processed using
     *  the \c AckReceive method. The \c Free method on the data buffer must
     *  also be invoked unless the \c PutBackReceivedData is used instead
```

I think we should not try to duplicate docs lest they become out of sync.

Reply:
OK, sounds good!  Could add a pointer to the TCPEndPoint documentation from here.